todo_app/test/test_client.py

```diff
@@ -0,0 +1,68 @@
+import pytest
+import requests
+import os
+from dotenv import load_dotenv, find_dotenv
+from todo_app import app
+
+
+@pytest.fixture
+def client():
+    # Use our test integration config instead of the 'real' version
+    file_path = find_dotenv('.env.test')
+    load_dotenv(file_path, override=True)
+
+    # Create the new app.
+    test_app = app.create_app()
+
+    # Use the app to create a test_client that can be used in our tests.
+    with test_app.test_client() as client:
+        yield client
+
+class StubResponse():
+    def __init__(self, fake_response_data):
+        self.fake_response_data = fake_response_data
+
+    def json(self):
+        return self.fake_response_data
+
+def stub(url, params={}):
+    test_board_id = os.environ.get('TRELLO_BOARD_ID')
+
+    if url == f'https://api.trello.com/1/boards/{test_board_id}/lists':
+        fake_response_data = [{
+            'id': '1a2b3c',
+            'name': 'To Do',
+            'cards': [{'id': '123', 'name': 'Test card 1'},{'id': '456', 'name': 'Test card 2'}]
+        }]
+        return StubResponse(fake_response_data)
+    elif url.startswith('https://api.trello.com/1/cards'):
+        fake_response_data = [{
+            'id': '1223def'
+        }]
+        return StubResponse(fake_response_data)
+
+    raise Exception(f'Integration test did not expect URL "{url}"')
+
+def test_index_page(monkeypatch, client):
+    # Replace requests.get(url) with our own function
+    monkeypatch.setattr(requests, 'get', stub)
+
+    # Make a request to our app's index page
+    response = client.get('/')
+
+    assert response.status_code == 200
+    assert '123' in response.data.decode()
+    assert 'Test card 2' in response.data.decode()
+
+def test_add_end_point(monkeypatch, client):
+    # Replace requests.post(url) with our own function
+    monkeypatch.setattr(requests, 'post', stub)
+
+    # Make a request to add a to do item
+    to_do_list_id = os.environ.get('TRELLO_TODO_LIST_ID')
+    payload = {'name': 'Test add item', 'idList': to_do_list_id}
+    
+    response = client.post('/add', data=payload)
+
+    assert response.status_code == 302
+    #assert 'Test add item' in response.data.decode()
```

Comment:
As you'll have seen, if we're using our reasonable but limited mocks, we are faking a response to the attempted call to Trello, but we don't actually add that data to our homepage. We _could_ extend our mock to handle that more thoroughly if we wanted to, e.g. something like:
* In the stub:
```python
created_items = []
def stub(url, params={}):
    test_board_id = os.environ.get('TRELLO_BOARD_ID')
    global created_items

    if url == f'https://api.trello.com/1/boards/{test_board_id}/lists':
        fake_response_data = [{
            'id': '1a2b3c',
            'name': 'To Do',
            'cards': [{'id': '123', 'name': 'Test card 1'},{'id': '456', 'name': 'Test card 2'}]
        }]
        # JACK: we also return any dynamically added items here
        fake_response_data[0]['cards'].extend(created_items)
        return StubResponse(fake_response_data)
    elif url.startswith('https://api.trello.com/1/cards'):
        fake_response_data = [{
            'id': '1223def'
        }]
        # JACK: Here we dynamically record the items being requested
        created_items.append({'id': '1223def', 'name': params['name']})
        return StubResponse(fake_response_data)

    raise Exception(f'Integration test did not expect URL "{url}"')
```

And we can then verify it by allowing this test to follow the redirects and check the new item appears on the right page:
```python
    response = client.post('/add', data=payload, follow_redirects = True)

    assert response.status_code == 200
    assert 'Test add item' in response.data.decode()
```